zproject/backends.py

```diff
@@ -285,22 +284,19 @@ def auth_complete(self, *args, **kwargs):
 
 class ZulipDummyBackend(ZulipAuthMixin):
     """
-    Used when we want to log you in but we don't know which backend to use.
+    Used when we want to log you in without checking any
+    authentication (i.e. new user registration or when otherwise
+    authentication has already been checked earlier in the process).
     """
 
-    def authenticate(self, username: Optional[Text]=None, realm_subdomain: Optional[Text]=None,
+    def authenticate(self, username: str=None, realm: Realm=None,
                      use_dummy_backend: bool=False,
-                     return_data: Optional[Dict[str, Any]]=None) -> Optional[UserProfile]:
-        assert username is not None
+                     return_data: Dict[str, Any]=None) -> Optional[UserProfile]:
         if use_dummy_backend:
-            user_profile = common_get_active_user_by_email(username)
-            if user_profile is None:
-                return None
-            if not user_matches_subdomain(realm_subdomain, user_profile):
-                if return_data is not None:
-                    return_data["invalid_subdomain"] = True
-                return None
-            return user_profile
+            # These are kwargs only for readability; they should never be None
+            assert username is not None
+            assert realm is not None
+            return common_get_active_user(username, realm, return_data)
```

Comment:
`return_data` is only used to provide nicer error messages -- whether or not `None` is returned is what controls whether auth happens.  So it's good that we're returning it, since it enables more cases to offer nice error messages for more cases; previously, the code paths that used this only had nice error messages for invalid_subdomain coming out of this (several of them effectively check `inactive_realm` earlier in their flow).